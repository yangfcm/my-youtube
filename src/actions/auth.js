import { SIGN_IN, SIGN_OUT, CATCH_ERROR } from "./types";
import { defAxios as axios } from "../settings";

export const signIn = (user) => {
  return async (dispatch) => {
    // console.log("action", user);
    try {
      if (!user) {
        throw Error("Authorization failed");
      }
      const accessToken = `${user.xc.token_type} ${user.xc.access_token}`;
      localStorage.setItem("access_token", accessToken);
      const response = await axios.get(
        "https://www.googleapis.com/oauth2/v1/userinfo?alt=json",
        {
          headers: {
            Authorization: accessToken,
          },
        }
      );
      dispatch({
        type: SIGN_IN,
        payload: {
          email: response.data.email,
          username: response.data.name,
          lastName: response.data.family_name,
          firstName: response.data.given_name,
          avatar: response.data.picture,
        },
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: "Authorization failed",
      });
    }
  };
};

export const signOut = () => {
  localStorage.removeItem("access_token");
  return {
    type: SIGN_OUT,
  };
};
