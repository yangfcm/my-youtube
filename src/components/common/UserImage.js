import React from "react";
import DropdownMenu from "./DropdownMenu";
import { mainMenuItems } from "../../settings";

class UserImage extends React.Component {
  state = {
    showDropdownMenu: false,
  };

  toggleDropdownMenu = () => {
    this.setState((prevState) => {
      return {
        showDropdownMenu: !prevState.showDropdownMenu,
      };
    });
  };

  closeDropdownMenu = () => {
    this.setState((prevState) => {
      if (prevState.showDropdownMenu) {
        return {
          showDropdownMenu: false,
        };
      }
    });
  };

  handleToggleDropdown = (e) => {
    this.toggleDropdownMenu();
    e.stopPropagation();
  };

  componentDidMount = () => {
    window.addEventListener("click", () => {
      this.closeDropdownMenu();
    });
  };

  render() {
    const { user } = this.props;
    if (!user) {
      return false;
    }
    return (
      <div style={{ position: "relative" }} className="mr-2">
        <div
          style={{ height: "55px", cursor: "pointer" }}
          id="dropdown"
          onClick={this.handleToggleDropdown}
        >
          {"      "}
          <img
            style={{ borderRadius: "50%", maxHeight: "100%" }}
            src={user.avatar}
            alt={user.username}
          />
        </div>
        <div
          id="dropdown-container"
          style={{
            position: "absolute",
            right: "0",
            zIndex: 100,
            minWidth: "150px",
            display: this.state.showDropdownMenu ? "block" : "none",
          }}
        >
          <DropdownMenu menuItems={mainMenuItems} />
        </div>
      </div>
    );
  }
}

export default UserImage;
