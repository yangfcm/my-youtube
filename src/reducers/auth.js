import { SIGN_IN, SIGN_OUT } from "../actions/types";

export const authReducer = (state = { signedIn: null, user: null }, action) => {
  switch (action.type) {
    case SIGN_IN:
      return {
        ...state,
        signedIn: true,
        user: action.payload,
      };
    case SIGN_OUT:
      return {
        ...state,
        signedIn: false,
        user: null,
      };
    default:
      return state;
  }
};
