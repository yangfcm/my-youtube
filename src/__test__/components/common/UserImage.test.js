import React from "react";
import { shallow } from "enzyme";
import UserImage from "components/common/UserImage";
import { authedUser } from "__test__/fixtures/auth";

describe("Test UserImage component", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<UserImage user={authedUser} />);
  });
  afterEach(() => {
    wrapper.unmount();
  });

  it("should render UserImage component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render correct image with user avatar", () => {
    expect(wrapper.find("img").prop("src")).toBe(authedUser.avatar);
  });

  it("can toggle dropdown menu by clicking user avatar", () => {
    const avatarDiv = wrapper.find("div#dropdown");
    expect(wrapper.find("div#dropdown-container").exists()).toBe(false);
    avatarDiv.prop("onClick")({ stopPropagation: jest.fn() });
    expect(wrapper.find("div#dropdown-container").exists()).toBe(true);
    avatarDiv.prop("onClick")({ stopPropagation: jest.fn() });
    expect(wrapper.find("div#dropdown-container").exists()).toBe(false);
  });
});
