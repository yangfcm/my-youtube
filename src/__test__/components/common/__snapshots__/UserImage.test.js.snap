// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test UserImage component can toggle dropdown menu by clicking user avatar 1`] = `
<div
  className="mr-2"
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    id="dropdown"
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
        "height": "55px",
      }
    }
  >
          
    <img
      alt="Test Test"
      src="https://lh3.googleusercontent.com/a-/AOh14GgHZ_IO2RUfg770gLs9V6UOsJaDMsWmGKP4AUna"
      style={
        Object {
          "borderRadius": "50%",
          "maxHeight": "100%",
        }
      }
    />
  </div>
  <div
    id="dropdown-container"
    style={
      Object {
        "minWidth": "150px",
        "position": "absolute",
        "right": "0",
        "zIndex": 100,
      }
    }
  >
    <Connect(GoogleAuthWrapper)
      menuItems={
        Array [
          Object {
            "exact": true,
            "icon": "thumbs-up",
            "link": "/",
            "name": "Recommend",
          },
          Object {
            "exact": false,
            "icon": "user-plus",
            "link": "/channel",
            "name": "Subscriptions",
          },
          Object {
            "exact": false,
            "icon": "list",
            "link": "/playlist",
            "name": "Play List",
          },
        ]
      }
    />
  </div>
</div>
`;

exports[`Test UserImage component should render UserImage component 1`] = `
<div
  className="mr-2"
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    id="dropdown"
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
        "height": "55px",
      }
    }
  >
          
    <img
      alt="Test Test"
      src="https://lh3.googleusercontent.com/a-/AOh14GgHZ_IO2RUfg770gLs9V6UOsJaDMsWmGKP4AUna"
      style={
        Object {
          "borderRadius": "50%",
          "maxHeight": "100%",
        }
      }
    />
  </div>
</div>
`;
